### PR TITLE
Add language option in settings menu

### PR DESCRIPTION
**Pull request by srifqi**
Rebase of #1937

### CODE REVIEW

**Comment by srifqi** [comment]
Okay, no merge :(
and get merge conflicts, again!
@SmallJoker please help me again!

**Comment by srifqi** [comment]
Okay, merge conflicts again. :(
Need code change for NEW layout by someone.

File: builtin/mainmenu/tab_settings.lua [hunk 1]
```diff
@@ -128,6 +128,31 @@ local function formspec(tabview, name, tabdata)
...
+	
+	for i=1, #locale_list, 1 do
+		-- check if translation file exists
+		if file_exists(
+			locale_path .. locale_list[i] .. DIR_DELIM
+			.. "LC_MESSAGES"
+			.. DIR_DELIM .. "minetest.mo"
+		) then
+			language_string = language_string .. "," .. locale_list[i]
+			language_count = language_count +1
```
Comment (nerzhul) [inline builtin/mainmenu/tab_settings.lua]: Space

**Comment by nerzhul** [comment]
If you rebase this could be good for merge. I agree with this change but i need to test it after rebase

**Comment by srifqi** [comment]
Okay, updated to fit new 0.4.12's layout.

PS: @SmallJoker please rebase this!

**Comment by srifqi** [comment]
Thanks, @SmallJoker for the rebase!

Also, please update the translation (aka `minetest.pot` aka `.po` files) after merge.
As new word in this pr, `Language:`

**Comment by rubenwardy** [comment]
+1 to idea. A check box in the top right may be better, just white text until you click it.

**Comment by srifqi** [comment]
@rubenwardy Sorry, what do you mean:
> A check box in the top right may be better, just white text until you click it.

**Comment by rubenwardy** [comment]
language

**Comment by srifqi** [comment]
Okay, @rubenwardy 
Thanks for ideas!
I'll update ASAP.

**Comment by Wuzzy2** [comment]
I am opposed to this feature unless it allows the user to select “System language” as well (which just clears the `language` setting in `minetest.conf`), and this must be pre-selected at the first start.
So that Minetest will still start automatically in the system language in a sanely-configured operating system.

**Comment by rubenwardy** [comment]
@Wuzzy, of course that will still be the case.

**Comment by Wuzzy2** [comment]
Good. :-)

**Comment by srifqi** [comment]
@Wuzzy2, another good idea!
Let me check how to do that, or maybe any thoughts?

**Comment by srifqi** [comment]
@Wuzzy2 @rubenwardy
Sorry, do not know how to that maybe in another PR.
So, this PR is final. Merge?

File: builtin/mainmenu/tab_settings.lua [hunk 2]
```diff
@@ -176,6 +176,31 @@ local function formspec(tabview, name, tabdata)
 		end
 	end
 
+	local language_formspec_string = "en" -- add english first because its default
+	local language_current_idx = 1
+	local language_count = 1
+	local language_selected = core.setting_get("language") or "en"
```
Comment (ShadowNinja) [inline builtin/mainmenu/tab_settings.lua]: This doesn't account for things like the `LANG` env var which `init_gettext` does.

Comment (srifqi) [inline builtin/mainmenu/tab_settings.lua]: @ShadowNinja How to account things like the `LANG` env var?

**Comment by srifqi** [comment]
Okay, @ShadowNinja
Updated, as your request.
Will be squashed soon.

**Comment by rubenwardy** [comment]
Is it possible that os.getevn("LANG") returns nil?

**Comment by srifqi** [comment]
@rubenwardy Haven't tested. Should I add `if` statement?

**Comment by Wuzzy2** [comment]
> Is it possible that os.getevn("LANG") returns nil?

No, this code does not even work. :P

If you meant `os.getenv("LANG")`, then the answer is “Yes.”. Because environment variables such as `LANG` can certainly be not defined at all. Next time, refer to the Lua documentation. :P
But in the case of `LANG`, it is most likely a sign of a poorly configured system if this environment variable is not set.

**Comment by srifqi** [comment]
Then, should I add `if` statement?
Like this:
```lua
local LANG_env = os.getenv("LANG")
if LANG_env  == nil or LANG_env  == "" then
	LANG_env = "en"
end
```

**Comment by rubenwardy** [comment]
Wuzzy, that was obviously a typo. I read the documentation today when the
author asked how to read environment variables in #minetest-mods. I don't
appreciate your tone.

You might be able to do: (pseudo code)

    setting or envLANG or "en"

Instead of an if statement, but I've never tried three in a row like that.

**Comment by Wuzzy2** [comment]
I am not sure if it is so simple by just reading `LANG` and applying the value directly, even with `nil` check. I am not very familar with locales and stuff, sorry. :-(

For example, on my system, `LANG` contains `de_DE.utf8`.

**Comment by mahmutelmas06** [comment]
Its better to restart game when language changed instead of showing a message.

**Comment by MoNTE48** [comment]
2015-09-11 08:21:12: ERROR[main]: ========== ERROR FROM LUA ===========
2015-09-11 08:21:12: ERROR[main]: Failed to load and run script from 
2015-09-11 08:21:12: ERROR[main]: C:\Users\monte\Desktop\minetest-0.4.13\bin\..\builtin\init.lua:
2015-09-11 08:21:12: ERROR[main]: ...minetest-0.4.13\bin\..\builtin\mainmenu\tab_settings.lua:196: attempt to call field 'get_dirlist' (a nil value)
2015-09-11 08:21:12: ERROR[main]: stack traceback:
2015-09-11 08:21:12: ERROR[main]: 	...minetest-0.4.13\bin\..\builtin\mainmenu\tab_settings.lua:196: in main chunk
2015-09-11 08:21:12: ERROR[main]: 	[C]: in function 'dofile'
2015-09-11 08:21:12: ERROR[main]: 	...Desktop\minetest-0.4.13\bin\..\builtin\mainmenu\init.lua:43: in main chunk
2015-09-11 08:21:12: ERROR[main]: 	[C]: in function 'dofile'
2015-09-11 08:21:12: ERROR[main]: 	...rs\monte\Desktop\minetest-0.4.13\bin\..\builtin\init.lua:31: in main chunk
2015-09-11 08:21:12: ERROR[main]: ======= END OF ERROR FROM LUA ========
2015-09-11 08:21:12: ERROR[main]: No future without mainmenu

**Comment by kilbith** [comment]
@PilzAdam You can close this.

**Comment by PilzAdam** [comment]
006ef5b4a5e362eed848ca26b411a8059d74b965
This is the 6th issue closed by this commit. I'm really productive :D

**Comment by srifqi** [comment]
Okay,
At least, thanks for responding here.

Off-topic: But, I a little bit don't like the tree-like settings, How about more dialog? ... and some presets?

**Pull request closed without merging**